### ISSUE TITLE
Make implications of `security.pam.yubico.enable` more obvious

### ISSUE THREAD

**Issue by MalteT** [NixOS/nixpkgs#166076]
### Describe the bug
Enabling `security.pam.yubico.enable` will allow the yubikey to be used for **every** PAM service that is not overriden via `security.pam.services.<name>.yubicoAuth = false`. I was quite astonished to be able to `ssh` into my machine without having to enter any password, just because the Yubikey was plugged into the target host.

### Steps To Reproduce
1. Set `security.pam.yubico.enable = true` in your NixOS config and configure the challenge-response cipher. `security.pam.yubico.control = "sufficient"` is probably required aswell.
2. Set `services.openssh.enable = true`.
3. Plug your Yubikey into your computer.
4. `ssh` into your machine from every other computer in your network.

### Expected behavior
This should not be the "default" behaviour. The problematic part in the above configuration is `security.pam.yubico.enable`.
In my opinion (without much knowledge on the topic) the option should be deprecated. Less extreme options include:

- Disabling `security.pam.services.sshd.yubicoAuth` by default.
- Throwing a warning when setting `security.pam.yubico.enable`

User @lvkm:matrix.org has also [mentioned their concerns about](https://matrix.to/#/!KqkRjyTEzAGRiZFBYT:nixos.org/$0HSeisSC8Oj68grIg16klAUtZTsxhLCNoeoYURgQdpk?via=nixos.org&via=matrix.org&via=tchncs.de) https://github.com/NixOS/nixpkgs/blob/1d08ea2bd83abef174fb43cbfb8a856b8ef2ce26/nixos/modules/services/networking/ssh/sshd.nix#L512

### Notify maintainers

@dtzWill (sorry for putting you here, it's been a long time since you've worked on `security.pam`)

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.16.10, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0pre20220127_558c4ee`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by mohe2015** [comment]
@MalteT For what pam service do you intend to use this? Can you tell me whether security.pam.services.<name>.yubicoAuth works instead. While I don't use this module I would think most people want this for logging in. I don't know if that is the "login" pam service. Would be great if you could test that and it would be awesome if you could create a PR that adds a deprecation message with hinting to the replacement if that's the case.

**Comment by MalteT** [comment]
Yes! I've actually managed to fix the hole with some help from the NixOS matrix server:

The complete config for reference:
```nix
 security.pam = {
    yubico = {
      enable = false;
      mode = "challenge-response";
      control = "sufficient";
    };
    services.login.yubicoAuth = true;
    services.swaylock.yubicoAuth = true;
  };
```

Replacing the problematic option with `security.pam.services.<name>.yubicoAuth = true` fixed the issue. I've tested logging in (both local and remotely) aswell as swaylock and everything seems to behave (no login via ssh without password entry/key usage).

I'll assemble a PR. Thanks for your quick response!

**Comment by MalteT** [comment]
Looking into this a bit more, deprecating the "global yubicoAuth" option will only scratch the surface. The following options are used as defaults within the `security.pam.services.<name>` module:
- `security.pam.p11.enable`
- `security.pam.u2f.enable`
- `security.pam.yubicoAuth.enable`
- `security.pam.usb.enable`
- `security.pam.enableOTPW`
- `services.fprintd.enable`
- `security.pam.oath.enable`
- (`security.pam.mount.enable`)

And while randomly using your fingerprint sensor is less likely than having your authentication token connected, enabling any of the above leads to (possibly) unwanted behaviour. I'm mostly concerned about the `sshd` pam service, but there might be other, less frequently used, services that could be affected. Note also, that `fprintd` is listed under `services` instead of `security`, which makes it even less obvious.

I know that there has at least been one attempt at rewriting the pam module and I'm not too eager to start another 😁 

Any ideas on how to handle this?

Also: without `security.pam.yubico.control = "sufficient"` no problems should occur (wrt yubicoAuth)

**Comment by mohe2015** [comment]
mis-clicked.

@MalteT Thanks for verifying that there is a good replacement. I think we could deprecate all of them and add a message for them with the replacement and some hints on which pam-services to enable. I quickly skimmed the module and it seems like most of these have a replacement. I think we could deprecate the remaining ones without a replacement for now if it's too hard to add the replacement.

**Comment by nix-config-storage** [comment]
> The following options are used as defaults within the security.pam.services.<name> module:
> 
> security.pam.p11.enable
> security.pam.u2f.enable
> security.pam.yubico.enable
> security.pam.usb.enable
> security.pam.enableOTPW
> services.fprintd.enable
> security.pam.oath.enable
> (security.pam.mount.enable)

Honestly, this seems like a very common/prevalent issue through ought Nixpkgs. There are a lot of modules that by default have options set to something, and in some cases it even seemingly causes the module itself to remain enabled if its ever evaluated
